src/common/FastCDC.cc

```diff
@@ -0,0 +1,158 @@
+    size_window_bits = SIZE_WINDOW_BITS;
+  }
+  min_bits = target - size_window_bits;
+  max_bits = target + size_window_bits;
+
+  std::mt19937_64 engine;
+
+  // prefill table
+  for (unsigned i = 0; i < 256; ++i) {
+    table[i] = engine();
+  }
+
+  // set mask
+  int did = 0;
+  uint64_t m = 0;
+  while (did < target_bits + TARGET_WINDOW_MASK_BITS) {
+    uint64_t bit = 1ull << (engine() & 63);
+    if (m & bit) {
+      continue;	// this bit is already set
+    }
+    m |= bit;
+    ++did;
+    if (did == target_bits - TARGET_WINDOW_MASK_BITS) {
+      large_mask = m;
+    } else if (did == target_bits) {
+      target_mask = m;
+    } else if (did == target_bits + TARGET_WINDOW_MASK_BITS) {
+      small_mask = m;
+    }
+  }
+}
+
+static inline bool _scan(
+  // these are our cursor/postion...
+  bufferlist::buffers_t::const_iterator *p,
+  const char **pp, const char **pe,
+  size_t& pos,
+  size_t max,   // how much to read
+  uint64_t& fp, // fingerprint
+  uint64_t mask, const uint64_t *table)
+{
+  while (pos < max) {
+    if (*pp == *pe) {
+      ++(*p);
+      *pp = (*p)->c_str();
+      *pe = *pp + (*p)->length();
+    }
+    const char *te = std::min(*pe, *pp + max - pos);
+    for (; *pp < te; ++(*pp), ++pos) {
+      if ((fp & mask) == mask) {
+	return false;
+      }
+      fp = (fp << 1) ^ table[*(unsigned char*)*pp];
+    }
+    if (pos >= max) {
+      return true;
+    }
+  }
+  return true;
+}
+
+void FastCDC::calc_chunks(
+  const bufferlist& bl,
+  std::vector<std::pair<uint64_t, uint64_t>> *chunks) const
+{
+  if (bl.length() == 0) {
+    return;
+  }
+  auto p = bl.buffers().begin();
+  const char *pp = p->c_str();
+  const char *pe = pp + p->length();
+
+  size_t pos = 0;
+  size_t len = bl.length();
+  while (pos < len) {
+    size_t cstart = pos;
+    uint64_t fp = 0;
+
+    // are we left with a min-sized (or smaller) chunk?
+    if (len - pos <= (1ul << min_bits)) {
+      chunks->push_back(std::pair<uint64_t,uint64_t>(pos, len - pos));
+      break;
+    }
+
+    // skip forward to the min chunk size cut point (minus the window, so
+    // we can initialize the rolling fingerprint).
+    size_t skip = (1 << min_bits) - window;
+    pos += skip;
+    while (skip) {
+      size_t s = std::min<size_t>(pe - pp, skip);
+      skip -= s;
+      pp += s;
+      if (pp == pe) {
+	++p;
+	pp = p->c_str();
+	pe = pp + p->length();
+      }
+    }
+    
+    // first fill the window
+    size_t max = pos + window;
+    while (pos < max) {
+      if (pp == pe) {
+	++p;
+	pp = p->c_str();
+	pe = pp + p->length();
+      }
+      const char *te = std::min(pe, pp + (max - pos));
+      for (; pp < te; ++pp, ++pos) {
+	fp = (fp << 1) ^ table[*(unsigned char*)pp];
+      }
+    }
+    ceph_assert(pos < len);
+
+    // find an end marker
+    if (_scan(&p, &pp, &pe, pos,  // for the first "small" region
+	      std::min(len,
+		       cstart + (1 << (target_bits - TARGET_WINDOW_BITS))),
+	      fp, small_mask, table) &&
```

Comment:
Just question. What is the purpose of this?
In the paper, 
>for ; i < NormalSize; i++; do
fp = (fp << 1) + Gear[ src[i] ];
if ! ( fp & MaskS ) then
return i; //if the masked bits are all ‘0’
for ; i<n; i++; do
fp = (fp << 1) + Gear[ src[i] ];
if ! ( fp & MaskL ) then
return i;

It seems that below two conditions ((target_bits + TARGET_WINDOW_BITS) and (1 << max_bits))
match both MaskS and MaskL..

Reply:
I'm not quite sure what you mean.

Reply:
I changed things a bit from the paper.  For the original FastCDC, anything < target_size uses maskS, and anything > uses maskL.  This implementation uses maskS < target_size/2, mask for target_size/2 to target_size*2, and maskL above that.

Reply:
Ok. It means the normalization level is 3?

Reply:
Normalization level is currently 2:
```
//  how many bits to set/clear in the small/large masks
#define TARGET_WINDOW_MASK_BITS  2
```

Reply:
I will update the source to describe how it corresponds to the paper's terms.

Reply:
I did a quick test (not super comprehensive, but something) comparing target_windwo_bits 1 vs 0 (==original fastcdc) and 1 performed slightly better.  Probably need to use a more representative data set to be sure, though.